src/Microsoft.OData.Core/SelectedPropertiesNode.cs

```diff
@@ -835,10 +835,30 @@ private static SelectedPropertiesNode ProcessSubExpand(string nodeName, Selected
         /// <returns>The generated SelectedPropertiesNode.</returns>
         private static SelectedPropertiesNode CombineSelectAndExpandResult(IEnumerable<string> selectList, IEnumerable<SelectedPropertiesNode> expandList)
         {
-            List<string> rawSelect = selectList.ToList();
-            rawSelect.RemoveAll(expandList.Select(m => m.nodeName).Contains);
+            HashSet<string> expandSet = new HashSet<string>();
+            bool isEntireSubTree = true;
 
-            if (rawSelect.Count == 0 && expandList.All(n => n.IsEntireSubtree()))
+            foreach(SelectedPropertiesNode propNode in expandList)
+            {
+                expandSet.Add(propNode.nodeName);
+
+                if (!propNode.IsEntireSubtree())
+                {
+                    isEntireSubTree = false;
+                }
+            }
+
+            List<string> rawSelect = new List<string>();
+
+            foreach(string name in selectList)
+            {
+                if (!expandSet.Contains(name))
+                {
+                    rawSelect.Add(name);
+                }
+            }
+                       
+            if (rawSelect.Count == 0 && isEntireSubTree)

```

Comment:
>rawSelect.Count == 0 &&  [](http://example.com/codeflow?start=16&length=24)

Rather than count the items in rawSelect to see if there are zero, just set isEntireSubTree to false between line 857 and 858 (as well as 847) and here just check isEntireSubTree.